src/proto/console/serial.rs

```diff
@@ -115,6 +120,17 @@ impl Serial {
     }
 }
 
+impl fmt::Write for Serial {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        let bytes = s.as_bytes();
+        if self.write(bytes).map_err(|_| fmt::Error)? != bytes.len() {
+            Err(fmt::Error)
```

Comment:
Unfortunately, fmt::Error does not give us a chance to differentiate timeouts from other kinds of failure.

Reply:
Unfortunately serial comm is really hard to get right, and while our simple test code works, I would feel bad to publicly expose this trait.

To begin with, we'd need to ensure `HARDWARE_FLOW_CONTROL` is enabled, otherwise just writing stuff to it isn't guaranteed to work. I think it would be better to write an abstraction only in the test runner crate, and use that for testing.

Reply:
Done. Since I don't need the general Write abstraction in multiple places, I just replaced it with its manual equivalent for now.